src/main/scala/schemas.scala

```diff
@@ -7,11 +7,11 @@ object ScalaSchema {
 }
 
 object JsonSchema {
-  type Prim[A] = JsonPrim[A]
+  type Prim[A] = JsonPrim
 
-  sealed trait JsonPrim[A]
-  final case class JsonString(value: String) extends JsonPrim[String]
-  final case class JsonNumber(value: BigDecimal) extends JsonPrim[BigDecimal]
-  final case class JsonBool(value: Boolean) extends JsonPrim[Boolean]
-  case object JsonNull extends JsonPrim[Nothing]
+  sealed trait JsonPrim
+  case object JsonString extends JsonPrim
+  case object JsonNumber extends JsonPrim
+  case object JsonBool extends JsonPrim
+  case object JsonNull extends JsonPrim
```

Comment:
I might suggest:

```scala
sealed trait JsonPrim[A]
final case object JsonString extends JsonPrim[String]
final case object JsonNumber extends JsonPrim[BigDecimal]
final case object JsonBool extends JsonPrim[Boolean]
final case object JsonNull extends JsonPrim[Null.type]
```

This may prevent casts down the road...

Reply:
@jdegoes cool, I can see that. What I don't completely understand is `final` for `case object`s. Is it some trickery for the compiler?

Reply:
@jdegoes I don't get it. 

Given the current definition of `SchemaModule#prim` what you suggest would prevent things like:

```scala
val localDateTimeSchema: Schema[LocalDateTime] = prim(JsonString)
```

Reply:
Because `JsonString` extends `JsonPrim[String]` (not `JsonPrim[A]` for arbitrary `A`), therefore we know that only 4 types of primitives can exist: `String`, `BigDecimal`, `Boolean`, `Null.type`. The compiler will not allow you to call `prim(JsonString)` and get back anything other than `Schema[String]`.

Reply:
Alright, I might be mixing things up there. But let me elaborate with a small example.

Suppose we have a `Message` class:

```scala
final case class Message(time: LocalDateTime, msg: String)
``` 

And we want to build a schema such as instances of that class are represented in JSON like so:

```json
{
  "time": "2018-08-14T17:42:15Z",
  "msg" : "This is Spartaz!!!"
}
```

We definitely need some (convenient) way to say that the type of the `time` field's value is `LocalDateTime` in the scala world and `string` in the JSON world.

Maybe a solution would be to add a combinator to `JsonPrim` that given a (partial) isomorphism between `A` and `B` (ie a pair of functions `A => Option[B]` and `B => A`) would transform a `JsonPrim[A]` into a `JsonPrim[B].

Reply:
Here's how we do that:

1. You have `Schema[String]` from `prim(JsonString)`.
2. You need `Schema[LocalDateTime]` from `Schema[String]`.
3. So you have to map over `prim(JsonString)` to obtain `Schema[LocalDateTime]`.

We only have to define `map` functions on `Schema`.

For example:

```scala
final case class SchemaMap[A, B](original: Schema[A], iso: Iso[A, B]) extends Schema[B]
```

However, like you say, `Iso` is too strong in general. We probably need `Codec[A, B]` that can fail in either direction.

With such a map schema, you can now call `.map` on the `prim(JsonString)` in order to transform it into `Schema[LocalDateTime]`, with possibility of failure in one direction.

Reply:
I originally thought we could do `imap` (I created a ticket for this), but I see it's not strong enough and we can't implement it for the primitives, although we can do so for other types. So adding a `SchemaTransformed` / `SchemaMap` node seems useful. 

The other way to do this might be as a fake sum type. Because then the `Prism` will allow you to fail on get (but not on set). But then you have the sum term id which serves no purpose for something like this and might be encoded poorly into JSON, etc. So introducing a map node with an explicit opportunity for failure seems like it might be better...